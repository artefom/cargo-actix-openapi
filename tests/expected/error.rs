//! API auto-generated by apigen

use std::fmt::Display;

use std::{collections::HashMap, fmt::Debug};

use serde::{Deserialize, Serialize};

use actix_web::{
    get, http::StatusCode, middleware::NormalizePath, web, App, HttpRequest, HttpResponse,
    HttpServer, ResponseError,
};

use async_trait::async_trait;

// Defaults
// -------------------------------


// Enums
// -------------------------------


// Struct
// -------------------------------

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct GreetUserPath {
    pub user: String,
}


// Error with details
// -------------------------------

/// Bails with detailed api error
#[macro_export]
macro_rules! apibail {
    ($err:expr,$msg:expr) => {
        return Err($crate::server::api::Detailed {
            error: $err,
            details: $msg.to_string(),
        })
    };
}

pub trait StatusCoded {
    fn status_code(&self) -> StatusCode;
}

#[derive(Debug)]
pub struct Detailed<E> {
    pub error: E,
    pub details: String,
}

impl<E: Display> Display for Detailed<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}. Reason: {}", self.error, self.details)
    }
}

impl<E: Display + Debug> std::error::Error for Detailed<E> {}

impl<E: Display + Debug> ResponseError for Detailed<E>
where
    E: StatusCoded,
{
    fn status_code(&self) -> StatusCode {
        self.error.status_code()
    }
}

/// Converts some result to Result with detailed api error
pub trait ApiErr<T, E> {
    /// Wrap the error value with additional context.
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static;
}

impl<T, E> ApiErr<T, E> for Result<T, E>
where
    E: Debug + Send + Sync + 'static,
{
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static,
    {
        // Not using map_err to save 2 useless frames off the captured backtrace
        // in ext_context.
        match self {
            Ok(ok) => Ok(ok),
            Err(original_error) => Err(Detailed {
                error: err,
                details: format!("{:?}", original_error),
            }),
        }
    }
}


// Error
// -------------------------------

/// Status NOT_FOUND:
/// User not found
/// 
/// Status BAD_REQUEST:
/// Input data error
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub enum GreetUserResponseError {
    NotFound,
    InvalidCharacterInName,
    NameContainsSpace,
}

impl Display for GreetUserResponseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NotFound => {
                write!(f, "Not found")
            },
            Self::InvalidCharacterInName => {
                write!(f, "Invalid character in name")
            },
            Self::NameContainsSpace => {
                write!(f, "Name contains space")
            },
        }
    }
}

impl StatusCoded for GreetUserResponseError {
    fn status_code(&self) -> StatusCode {
        match self {
            Self::NotFound => StatusCode::NOT_FOUND,
            Self::InvalidCharacterInName => StatusCode::BAD_REQUEST,
            Self::NameContainsSpace => StatusCode::BAD_REQUEST,
        }
    }
}
