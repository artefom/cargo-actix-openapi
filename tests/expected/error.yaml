api:
  definitions:
    DOCS_OPENAPI:
      data: !StaticStr
        path: static/openapi.yaml
    DOCS_HTML:
      data: !StaticStr
        path: static/docs.html
    openapi:
      data: !StaticStringPath
        data: DOCS_OPENAPI
    docs:
      data: !StaticHtmlPath
        data: DOCS_HTML
    to_docs:
      data: !Redirect
        target: v1/docs
    GreetUserPath:
      data: !Struct
        doc: null
        properties:
        - name: user
          rename: user
          default: null
          type_: String
          doc: The name of the user to greet.
    GreetUserError:
      data: !ApiErr
        doc: |-
          Status NOT_FOUND:
          User not found

          Status BAD_REQUEST:
          Input data error
        variants:
        - name: NotFound
          detail: Not found
          code: NOT_FOUND
        - name: InvalidCharacterInName
          detail: Invalid character in name
          code: BAD_REQUEST
        - name: NameContainsSpace
          detail: Name contains space
          code: BAD_REQUEST
  operations:
    greet_user:
      doc: Returns a greeting to the user!
      param_path: web::Path<GreetUserPath>
      param_query: null
      param_body: null
      response: Result<web::Json<String>,Detailed<GreetUserError>>
  paths:
  - operation: greet_user
    path: /hello/{user}
    method: Get
  - operation: greet_user
    path: /v1/hello/{user}
    method: Get
  static_services:
  - method: Get
    path: /openapi.yaml
    data: openapi
  - method: Get
    path: /docs
    data: docs
  - method: Get
    path: /v1
    data: to_docs
  - method: Get
    path: /v1/openapi.yaml
    data: openapi
  - method: Get
    path: /v1/docs
    data: docs
